internal/buffer/save.go

```diff
@@ -178,60 +298,77 @@ func (b *Buffer) saveToFile(filename string, withSudo bool, autoSave bool) error
 		}
 	}
 
-	var fileSize int
-
-	enc, err := htmlindex.Get(b.Settings["encoding"].(string))
+	saveResponseChan := make(chan saveResponse)
+	saveRequestChan <- saveRequest{b, absFilename, withSudo, saveResponseChan}
+	result := <-saveResponseChan
+	err = result.err
 	if err != nil {
-		return err
-	}
-
-	fwriter := func(file io.Writer) (e error) {
-		if len(b.lines) == 0 {
-			return
-		}
-
-		// end of line
-		var eol []byte
-		if b.Endings == FFDos {
-			eol = []byte{'\r', '\n'}
-		} else {
-			eol = []byte{'\n'}
-		}
-
-		// write lines
-		if fileSize, e = file.Write(b.lines[0].data); e != nil {
-			return
-		}
-
-		for _, l := range b.lines[1:] {
-			if _, e = file.Write(eol); e != nil {
-				return
-			}
-			if _, e = file.Write(l.data); e != nil {
-				return
-			}
-			fileSize += len(eol) + len(l.data)
+		if errors.Is(err, util.ErrOverwrite) {
+			screen.TermMessage(err)
+			err = errors.Unwrap(err)
 		}
-		return
-	}
-
-	if err = overwriteFile(absFilename, enc, fwriter, withSudo); err != nil {
 		return err
 	}
 
 	if !b.Settings["fastdirty"].(bool) {
-		if fileSize > LargeFileThreshold {
+		if result.size > LargeFileThreshold {
 			// For large files 'fastdirty' needs to be on
 			b.Settings["fastdirty"] = true
 		} else {
 			calcHash(b, &b.origHash)
 		}
 	}
 
-	b.Path = filename
-	absPath, _ := filepath.Abs(filename)
-	b.AbsPath = absPath
+	b.Path, _ = util.ReplaceHome(filename)
```

Comment:
I do not know much about the specific state of the pull request right now, but it seems like `Buffer.saveToFile` is still supposed to handle paths with tilde like before. Isn't the expanded path also supposed to be used instead in other lines referencing `filename`?

Reply:
This function uses `absFilename` (represented in `b.AbsPath` too) to store the file. The `b.Path` is used to fill the infobar with the files path.

Reply:
Sorry, I was not specific on what I meant. There are few times I execute the `Save` action on a non-empty buffer with no path, then enter a path like `~/desktop/test.txt` in the prompt. The action eventually calls `b.SaveAs` then `b.saveToFile`, but the path with tilde would be used to store the file with this pull request currently. I think it is a bug because `absFilename` should be set as `/home/user/desktop.txt`.

Reply:
Sorry, I was not specific on what I meant. There are few times I execute the `Save` action on a non-empty buffer with no path, then enter a path like `~/desktop/test.txt` in the prompt. The action eventually calls `b.SaveAs` then `b.saveToFile`, but the path with tilde would be used to store the file with this pull request currently. I think it is a bug because `absFilename` should be set as `/home/user/desktop.txt`.

Reply:
Thanks for reporting...I will fix it.

Reply:
Good catch.

> The `b.Path` is used to fill the infobar with the files path.

And there is actually an inconsistency (not caused by this PR): if we open a file `~/foo/bar`, the infobar shows its expanded path, i.e. `/home/user/foo/bar`, while if we save a file as `~/foo/bar`, the infobar shows it with tilde, i.e. `~/foo/bar`.

Reply:
The function has been updated now, but I think `filename` should be replaced with `absFilename` in `util.GetTime(filename)` (line 267) and `os.Stat(filename)` (line 271) too.

> > The `b.Path` is used to fill the infobar with the files path.
> 
> And there is actually an inconsistency (not caused by this PR): if we open a file `~/foo/bar`, the infobar shows its expanded path, i.e. `/home/user/foo/bar`, while if we save a file as `~/foo/bar`, the infobar shows it with tilde, i.e. `~/foo/bar`.

Yes, I think it has been fixed in this PR.

Reply:
> Yes, I think it has been fixed in this PR.

I don't think so.

Reply:
Ah sorry, I did not look properly. `b.Path` was assigned with the expanded path when I posted my comment about path with tilde being used, but it is now assigned with that path (`filename`) again.

Reply:
> And there is actually an inconsistency [...]

Argh, since `micro` may receive it already expanded by the calling shell it seems to be better to keep the `Path` with the expanded `~` as home directory...even I don't like it.

Reply:
> The function has been updated now, but I think `filename` should be replaced with `absFilename` in `util.GetTime(filename)` (line 267) and `os.Stat(filename)` (line 271).

Agree, or alternatively, on top of the latest version of the PR, move `filename, _ = util.ReplaceHome(filename)` before the first use of `filename`.

There is a handful of other places throughout micro where a file is opened, stat()ed etc using `b.Path` instead of `b.AbsPath`... Generally, as long as the path has been tilde-expanded, using both absolute and relative paths is correct, in particular even if the current directory is changed via the `cd` command, `b.Path` is updated accordingly (`CdCmd()` takes care of that). Except that, the current directory could be also changed via a plugin (and we even deliberately export `os.Chdir` to lua), goddamn...

Reply:
> goddamn...

Yep, it's currently kind of a mess when it comes to `.Path` vs `.AbsPath`.
This could be something for the future to clearly differentiate between them both, but this would be a breaking change (e.g. `.Path` -> `.RelPath`). :thinking: